// oazapfts/generate.js
import "lodash";
import "typescript";
import { Y as r, X as i, H as o, m as n, l as c, W as l, I as p, D as d, K as m, L as u, y as f, J as g, P as y, B as N, u as T, x as F, M as b, E as D, Q as P, F as O, N as h, G as C, O as I, n as k, R as x, z as A, k as M, A as j, a0 as w, V as B, S as E, g as S, U as V, b as L, T as Q, c as R, d as U, h as q, a as v, i as z, e as J, f as K, $ as G, o as H, w as W, p as X, Z as Y, _ as Z, q as _, r as $, s as ee, C as ae, j as te, v as se } from "./tscodegen-CQ_2EobI.js";
export {
  r as addComment,
  i as appendNodes,
  o as block,
  n as callOazapftsFunction,
  c as callQsFunction,
  l as changePropertyValue,
  p as createArrowFunction,
  d as createCall,
  m as createClassDeclaration,
  u as createConstructor,
  f as createEnumTypeNode,
  g as createFunctionDeclaration,
  y as createIndexSignature,
  N as createInterfaceAliasDeclaration,
  T as createKeywordType,
  F as createLiteral,
  b as createMethod,
  D as createMethodCall,
  P as createObjectBinding,
  O as createObjectLiteral,
  h as createParameter,
  C as createPropertyAssignment,
  I as createPropertySignature,
  k as createQuestionToken,
  x as createTemplateString,
  A as createTypeAliasDeclaration,
  M as createUrlExpression,
  j as default,
  w as defaultBaseUrl,
  B as findFirstVariableDeclaration,
  E as findNode,
  S as getBodyFormatter,
  V as getFirstDeclarationName,
  L as getFormatter,
  Q as getName,
  R as getOperationIdentifier,
  U as getOperationName,
  q as getReferenceName,
  v as isJsonMimeType,
  z as isMimeType,
  J as isNullable,
  K as isReference,
  G as isValidIdentifier,
  H as keywordType,
  W as modifier,
  X as printFile,
  Y as printNode,
  Z as printNodes,
  _ as questionToken,
  $ as refPathToPropertyPath,
  ee as supportDeepObjects,
  ae as toExpression,
  te as toIdentifier,
  se as verbs
};
