// @rtk-query/codegen-openapi/lib/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  generateEndpoints: () => generateEndpoints,
  parseConfig: () => parseConfig
});
module.exports = __toCommonJS(src_exports);
var import_node_fs = __toESM(require("node:fs"));
var import_node_module = require("node:module");
var import_node_path3 = __toESM(require("node:path"));

// src/generate.ts
var import_lodash = __toESM(require("lodash.camelcase"));
var import_node_path2 = __toESM(require("node:path"));
var import_generate3 = __toESM(require("oazapfts/generate"));
var import_typescript4 = __toESM(require("typescript"));

// src/utils/factory.ts
var import_typescript = __toESM(require("typescript"));
var import_semver = __toESM(require("semver"));
var originalFactory = import_typescript.default.factory;
function createImportSpecifier(propertyName, name) {
  if (import_semver.default.satisfies(import_typescript.default.version, ">= 4.5"))
    return originalFactory.createImportSpecifier(false, propertyName, name);
  return originalFactory.createImportSpecifier(propertyName, name);
}
function createExportSpecifier(propertyName, name) {
  if (import_semver.default.satisfies(import_typescript.default.version, ">= 4.5"))
    return originalFactory.createExportSpecifier(false, propertyName, name);
  return originalFactory.createExportSpecifier(propertyName, name);
}
var factory = {
  ...originalFactory,
  createImportSpecifier,
  createExportSpecifier
};

// src/codegen.ts
var import_typescript2 = __toESM(require("typescript"));
var defaultEndpointBuilder = factory.createIdentifier("build");
function generateObjectProperties(obj) {
  return Object.entries(obj).filter(([_, v]) => v).map(([k, v]) => factory.createPropertyAssignment(factory.createIdentifier(k), v));
}
function generateImportNode(pkg, namedImports, defaultImportName) {
  return factory.createImportDeclaration(
    void 0,
    factory.createImportClause(
      false,
      defaultImportName !== void 0 ? factory.createIdentifier(defaultImportName) : void 0,
      factory.createNamedImports(
        Object.entries(namedImports).map(
          ([propertyName, name]) => factory.createImportSpecifier(
            name === propertyName ? void 0 : factory.createIdentifier(propertyName),
            factory.createIdentifier(name)
          )
        )
      )
    ),
    factory.createStringLiteral(pkg)
  );
}
function generateCreateApiCall({
  endpointBuilder = defaultEndpointBuilder,
  endpointDefinitions,
  tag
}) {
  const injectEndpointsObjectLiteralExpression = factory.createObjectLiteralExpression(
    generateObjectProperties({
      endpoints: factory.createArrowFunction(
        void 0,
        void 0,
        [factory.createParameterDeclaration(void 0, void 0, endpointBuilder, void 0, void 0, void 0)],
        void 0,
        factory.createToken(import_typescript2.default.SyntaxKind.EqualsGreaterThanToken),
        factory.createParenthesizedExpression(endpointDefinitions)
      ),
      overrideExisting: factory.createFalse()
    }),
    true
  );
  if (tag) {
    const enhanceEndpointsObjectLiteralExpression = factory.createObjectLiteralExpression(
      [factory.createShorthandPropertyAssignment(factory.createIdentifier("addTagTypes"), void 0)],
      true
    );
    return factory.createVariableStatement(
      void 0,
      factory.createVariableDeclarationList(
        [
          factory.createVariableDeclaration(
            factory.createIdentifier("injectedRtkApi"),
            void 0,
            void 0,
            factory.createCallExpression(
              factory.createPropertyAccessExpression(
                factory.createCallExpression(
                  factory.createPropertyAccessExpression(
                    factory.createIdentifier("api"),
                    factory.createIdentifier("enhanceEndpoints")
                  ),
                  void 0,
                  [enhanceEndpointsObjectLiteralExpression]
                ),
                factory.createIdentifier("injectEndpoints")
              ),
              void 0,
              [injectEndpointsObjectLiteralExpression]
            )
          )
        ],
        import_typescript2.default.NodeFlags.Const
      )
    );
  }
  return factory.createVariableStatement(
    void 0,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createIdentifier("injectedRtkApi"),
          void 0,
          void 0,
          factory.createCallExpression(
            factory.createPropertyAccessExpression(
              factory.createIdentifier("api"),
              factory.createIdentifier("injectEndpoints")
            ),
            void 0,
            [injectEndpointsObjectLiteralExpression]
          )
        )
      ],
      import_typescript2.default.NodeFlags.Const
    )
  );
}
function generateEndpointDefinition({
  operationName,
  type,
  Response,
  QueryArg,
  queryFn,
  endpointBuilder = defaultEndpointBuilder,
  extraEndpointsProps,
  tags,
  tagOverrides
}) {
  const objectProperties = generateObjectProperties({ query: queryFn, ...extraEndpointsProps });
  const providesTags = tagOverrides && "providesTags" in tagOverrides ? tagOverrides.providesTags : type === "query" ? tags : void 0;
  const invalidatesTags = tagOverrides && "invalidatesTags" in tagOverrides ? tagOverrides.invalidatesTags : type === "mutation" ? tags : void 0;
  if (providesTags !== void 0) {
    objectProperties.push(
      factory.createPropertyAssignment(
        factory.createIdentifier("providesTags"),
        factory.createArrayLiteralExpression(providesTags.map((tag) => factory.createStringLiteral(tag)), false)
      )
    );
  }
  if (invalidatesTags !== void 0) {
    objectProperties.push(
      factory.createPropertyAssignment(
        factory.createIdentifier("invalidatesTags"),
        factory.createArrayLiteralExpression(invalidatesTags.map((tag) => factory.createStringLiteral(tag)), false)
      )
    );
  }
  return factory.createPropertyAssignment(
    factory.createIdentifier(operationName),
    factory.createCallExpression(
      factory.createPropertyAccessExpression(endpointBuilder, factory.createIdentifier(type)),
      [Response, QueryArg],
      [factory.createObjectLiteralExpression(objectProperties, true)]
    )
  );
}
function generateTagTypes({ addTagTypes }) {
  return factory.createVariableStatement(
    [factory.createModifier(import_typescript2.default.SyntaxKind.ExportKeyword)],
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createIdentifier("addTagTypes"),
          void 0,
          void 0,
          factory.createAsExpression(
            factory.createArrayLiteralExpression(
              addTagTypes.map((tagType) => factory.createStringLiteral(tagType)),
              true
            ),
            factory.createTypeReferenceNode(factory.createIdentifier("const"), void 0)
          )
        )
      ],
      import_typescript2.default.NodeFlags.Const
    )
  );
}

// src/generators/react-hooks.ts
var import_typescript3 = __toESM(require("typescript"));
var import_generate = require("oazapfts/generate");

// src/utils/capitalize.ts
function capitalize(str) {
  return str.replace(str[0], str[0].toUpperCase());
}

// src/types.ts
var operationKeys = ["get", "put", "post", "delete", "options", "head", "patch", "trace"];

// src/utils/getOperationDefinitions.ts
function getOperationDefinitions(v3Doc) {
  return Object.entries(v3Doc.paths).flatMap(
    ([path4, pathItem]) => !pathItem ? [] : Object.entries(pathItem).filter(
      (arg) => operationKeys.includes(arg[0])
    ).map(([verb, operation]) => ({
      path: path4,
      verb,
      pathItem,
      operation
    }))
  );
}

// src/utils/getV3Doc.ts
var import_swagger_parser = __toESM(require("@apidevtools/swagger-parser"));
var import_swagger2openapi = __toESM(require("swagger2openapi"));
async function getV3Doc(spec, httpResolverOptions) {
  const doc = await import_swagger_parser.default.bundle(spec, {
    resolve: {
      http: httpResolverOptions
    }
  });
  const isOpenApiV3 = "openapi" in doc && doc.openapi.startsWith("3");
  if (isOpenApiV3) {
    return doc;
  } else {
    const result = await import_swagger2openapi.default.convertObj(doc, {});
    return result.openapi;
  }
}

// src/utils/isQuery.ts
function isQuery(verb, overrides) {
  if (overrides?.type) {
    return overrides.type === "query";
  }
  return verb === "get";
}

// src/utils/isValidUrl.ts
function isValidUrl(string) {
  try {
    new URL(string);
  } catch (_) {
    return false;
  }
  return true;
}

// src/utils/prettier.ts
var import_node_path = __toESM(require("node:path"));
var import_prettier = __toESM(require("prettier"));
var EXTENSION_TO_PARSER = {
  ts: "typescript",
  tsx: "typescript",
  js: "babel",
  jsx: "babel",
  "js.flow": "flow",
  flow: "flow",
  gql: "graphql",
  graphql: "graphql",
  css: "scss",
  scss: "scss",
  less: "scss",
  stylus: "scss",
  markdown: "markdown",
  md: "markdown",
  json: "json"
};
async function prettify(filePath, content, prettierConfigFile) {
  let config = null;
  let parser = "typescript";
  if (filePath) {
    const fileExtension = import_node_path.default.extname(filePath).slice(1);
    parser = EXTENSION_TO_PARSER[fileExtension];
    config = await import_prettier.default.resolveConfig(process.cwd(), {
      useCache: true,
      editorconfig: !prettierConfigFile,
      config: prettierConfigFile
    });
  } else if (prettierConfigFile) {
    config = await import_prettier.default.resolveConfig(process.cwd(), {
      useCache: true,
      config: prettierConfigFile
    });
  }
  return import_prettier.default.format(content, {
    parser,
    ...config
  });
}

// src/utils/removeUndefined.ts
function removeUndefined(t) {
  return typeof t !== "undefined";
}

// src/generators/react-hooks.ts
var createBinding = ({
  operationDefinition: { verb, path: path4, operation },
  overrides,
  isLazy = false,
  operationNameSuffix
}) => factory.createBindingElement(
  void 0,
  void 0,
  factory.createIdentifier(
    `use${isLazy ? "Lazy" : ""}${capitalize((0, import_generate.getOperationName)(verb, path4, operation.operationId))}${operationNameSuffix ?? ""}${isQuery(verb, overrides) ? "Query" : "Mutation"}`
  ),
  void 0
);
var getReactHookName = ({ operationDefinition, endpointOverrides, config, operationNameSuffix }) => {
  const overrides = getOverrides(operationDefinition, endpointOverrides);
  const baseParams = {
    operationDefinition,
    overrides,
    operationNameSuffix
  };
  const _isQuery = isQuery(operationDefinition.verb, overrides);
  if (typeof config === "boolean") {
    return createBinding(baseParams);
  }
  if (_isQuery) {
    return [
      ...config.queries ? [createBinding(baseParams)] : [],
      ...config.lazyQueries ? [createBinding({ ...baseParams, isLazy: true })] : []
    ];
  }
  return config.mutations ? createBinding(baseParams) : [];
};
var generateReactHooks = ({
  exportName,
  operationDefinitions,
  endpointOverrides,
  config,
  operationNameSuffix
}) => factory.createVariableStatement(
  [factory.createModifier(import_typescript3.default.SyntaxKind.ExportKeyword)],
  factory.createVariableDeclarationList(
    [
      factory.createVariableDeclaration(
        factory.createObjectBindingPattern(
          operationDefinitions.map((operationDefinition) => getReactHookName({ operationDefinition, endpointOverrides, config, operationNameSuffix })).flat()
        ),
        void 0,
        void 0,
        factory.createIdentifier(exportName)
      )
    ],
    import_typescript3.default.NodeFlags.Const
  )
);

// src/generate.ts
var generatedApiName = "injectedRtkApi";
var v3DocCache = {};
function defaultIsDataResponse(code, includeDefault) {
  if (includeDefault && code === "default") {
    return true;
  }
  const parsedCode = Number(code);
  return !Number.isNaN(parsedCode) && parsedCode >= 200 && parsedCode < 300;
}
function getOperationName2({ verb, path: path4, operation }) {
  return (0, import_generate3.getOperationName)(verb, path4, operation.operationId);
}
function getTags({ verb, pathItem }) {
  return verb ? pathItem[verb]?.tags || [] : [];
}
function patternMatches(pattern) {
  const filters = Array.isArray(pattern) ? pattern : [pattern];
  return function matcher(operationName) {
    if (!pattern) return true;
    return filters.some(
      (filter) => typeof filter === "string" ? filter === operationName : filter?.test(operationName)
    );
  };
}
function operationMatches(pattern) {
  const checkMatch = typeof pattern === "function" ? pattern : patternMatches(pattern);
  return function matcher(operationDefinition) {
    if (!pattern) return true;
    const operationName = getOperationName2(operationDefinition);
    return checkMatch(operationName, operationDefinition);
  };
}
function argumentMatches(pattern) {
  const checkMatch = typeof pattern === "function" ? pattern : patternMatches(pattern);
  return function matcher(argumentDefinition) {
    if (!pattern || argumentDefinition.in === "path") return true;
    const argumentName = argumentDefinition.name;
    return checkMatch(argumentName, argumentDefinition);
  };
}
function withQueryComment(node, def, hasTrailingNewLine) {
  const comment = def.origin === "param" ? def.param.description : def.body.description;
  if (comment) {
    return import_typescript4.default.addSyntheticLeadingComment(
      node,
      import_typescript4.default.SyntaxKind.MultiLineCommentTrivia,
      `* ${comment} `,
      hasTrailingNewLine
    );
  }
  return node;
}
function getPatternFromProperty(property, apiGen) {
  const resolved = apiGen.resolve(property);
  if (!resolved || typeof resolved !== "object" || !("pattern" in resolved)) return null;
  if (resolved.type !== "string") return null;
  const pattern = resolved.pattern;
  return typeof pattern === "string" && pattern.length > 0 ? pattern : null;
}
function generateRegexConstantsForType(typeName, schema, apiGen) {
  const resolvedSchema = apiGen.resolve(schema);
  if (!resolvedSchema || !("properties" in resolvedSchema) || !resolvedSchema.properties) return [];
  const constants = [];
  for (const [propertyName, property] of Object.entries(resolvedSchema.properties)) {
    const pattern = getPatternFromProperty(property, apiGen);
    if (!pattern) continue;
    const constantName = (0, import_lodash.default)(`${typeName} ${propertyName} Pattern`);
    const escapedPattern = pattern.replaceAll("/", String.raw`\/`);
    const regexLiteral = factory.createRegularExpressionLiteral(`/${escapedPattern}/`);
    constants.push(
      factory.createVariableStatement(
        [factory.createModifier(import_typescript4.default.SyntaxKind.ExportKeyword)],
        factory.createVariableDeclarationList(
          [
            factory.createVariableDeclaration(
              factory.createIdentifier(constantName),
              void 0,
              void 0,
              regexLiteral
            )
          ],
          import_typescript4.default.NodeFlags.Const
        )
      )
    );
  }
  return constants;
}
function getOverrides(operation, endpointOverrides) {
  return endpointOverrides?.find((override) => operationMatches(override.pattern)(operation));
}
async function generateApi(spec, {
  apiFile,
  apiImport = "api",
  exportName = "enhancedApi",
  argSuffix = "ApiArg",
  responseSuffix = "ApiResponse",
  operationNameSuffix = "",
  hooks = false,
  tag = false,
  outputFile,
  isDataResponse = defaultIsDataResponse,
  filterEndpoints,
  endpointOverrides,
  unionUndefined,
  encodePathParams = false,
  encodeQueryParams = false,
  flattenArg = false,
  includeDefault = false,
  useEnumType = false,
  mergeReadWriteOnly = false,
  httpResolverOptions,
  useUnknown = false,
  esmExtensions = false,
  outputRegexConstants = false
}) {
  const v3Doc = v3DocCache[spec] ??= await getV3Doc(spec, httpResolverOptions);
  const apiGen = new import_generate3.default(v3Doc, {
    unionUndefined,
    useEnumType,
    mergeReadWriteOnly,
    useUnknown
  });
  if (apiGen.spec.components?.schemas) {
    apiGen.preprocessComponents(apiGen.spec.components.schemas);
  }
  const operationDefinitions = getOperationDefinitions(v3Doc).filter(operationMatches(filterEndpoints));
  const resultFile = import_typescript4.default.createSourceFile(
    "someFileName.ts",
    "",
    import_typescript4.default.ScriptTarget.Latest,
    /*setParentNodes*/
    false,
    import_typescript4.default.ScriptKind.TS
  );
  const printer = import_typescript4.default.createPrinter({ newLine: import_typescript4.default.NewLineKind.LineFeed });
  const interfaces = {};
  function registerInterface(declaration) {
    const name = declaration.name.escapedText.toString();
    if (name in interfaces) {
      throw new Error(`interface/type alias ${name} already registered`);
    }
    interfaces[name] = declaration;
    return declaration;
  }
  if (outputFile) {
    outputFile = import_node_path2.default.resolve(process.cwd(), outputFile);
    if (apiFile.startsWith(".")) {
      apiFile = import_node_path2.default.relative(import_node_path2.default.dirname(outputFile), apiFile);
      apiFile = apiFile.replace(/\\/g, "/");
      if (!apiFile.startsWith(".")) apiFile = `./${apiFile}`;
    }
  }
  if (esmExtensions === true) {
    apiFile = apiFile.replace(/\.mts$/, ".mjs").replace(/\.[jt]sx$/, ".jsx").replace(/\.ts$/, ".js");
  } else {
    apiFile = apiFile.replace(/\.[jt]sx?$/, "");
  }
  return printer.printNode(
    import_typescript4.default.EmitHint.Unspecified,
    factory.createSourceFile(
      [
        generateImportNode(apiFile, { [apiImport]: "api" }),
        ...tag ? [generateTagTypes({ addTagTypes: extractAllTagTypes({ operationDefinitions }) })] : [],
        generateCreateApiCall({
          tag,
          endpointDefinitions: factory.createObjectLiteralExpression(
            operationDefinitions.map(
              (operationDefinition) => generateEndpoint({
                operationDefinition,
                overrides: getOverrides(operationDefinition, endpointOverrides)
              })
            ),
            true
          )
        }),
        factory.createExportDeclaration(
          void 0,
          false,
          factory.createNamedExports([
            factory.createExportSpecifier(
              factory.createIdentifier(generatedApiName),
              factory.createIdentifier(exportName)
            )
          ]),
          void 0
        ),
        ...Object.values(interfaces),
        ...outputRegexConstants ? apiGen.aliases.flatMap((alias) => {
          if (!import_typescript4.default.isInterfaceDeclaration(alias) && !import_typescript4.default.isTypeAliasDeclaration(alias)) return [alias];
          const typeName = alias.name.escapedText.toString();
          const schema = v3Doc.components?.schemas?.[typeName];
          if (!schema) return [alias];
          const regexConstants = generateRegexConstantsForType(typeName, schema, apiGen);
          return regexConstants.length > 0 ? [alias, ...regexConstants] : [alias];
        }) : apiGen.aliases,
        ...apiGen.enumAliases,
        ...hooks ? [
          generateReactHooks({
            exportName: generatedApiName,
            operationDefinitions,
            endpointOverrides,
            config: hooks,
            operationNameSuffix
          })
        ] : []
      ],
      factory.createToken(import_typescript4.default.SyntaxKind.EndOfFileToken),
      import_typescript4.default.NodeFlags.None
    ),
    resultFile
  );
  function extractAllTagTypes({ operationDefinitions: operationDefinitions2 }) {
    const allTagTypes = /* @__PURE__ */ new Set();
    for (const operationDefinition of operationDefinitions2) {
      const { verb, pathItem } = operationDefinition;
      for (const tag2 of getTags({ verb, pathItem })) {
        allTagTypes.add(tag2);
      }
    }
    return [...allTagTypes];
  }
  function generateEndpoint({
    operationDefinition,
    overrides
  }) {
    const {
      verb,
      path: path4,
      pathItem,
      operation,
      operation: { responses, requestBody }
    } = operationDefinition;
    const operationName = getOperationName2({ verb, path: path4, operation });
    const tags = tag ? getTags({ verb, pathItem }) : void 0;
    const isQuery2 = isQuery(verb, overrides);
    const returnsJson = apiGen.getResponseType(responses) === "json";
    let ResponseType = factory.createKeywordTypeNode(import_typescript4.default.SyntaxKind.UnknownKeyword);
    if (returnsJson) {
      const returnTypes = Object.entries(responses || {}).map(
        ([code, response]) => [
          code,
          apiGen.resolve(response),
          apiGen.getTypeFromResponse(response, "readOnly") || factory.createKeywordTypeNode(import_typescript4.default.SyntaxKind.UndefinedKeyword)
        ]
      ).filter(
        ([status, response]) => isDataResponse(status, includeDefault, apiGen.resolve(response), responses || {})
      ).filter(([_1, _2, type]) => type !== import_generate3.keywordType.void).map(
        ([code, response, type]) => import_typescript4.default.addSyntheticLeadingComment(
          { ...type },
          import_typescript4.default.SyntaxKind.MultiLineCommentTrivia,
          `* status ${code} ${response.description} `,
          false
        )
      );
      if (returnTypes.length > 0) {
        ResponseType = factory.createUnionTypeNode(returnTypes);
      }
    }
    const ResponseTypeName = factory.createTypeReferenceNode(
      registerInterface(
        factory.createTypeAliasDeclaration(
          [factory.createModifier(import_typescript4.default.SyntaxKind.ExportKeyword)],
          capitalize(operationName + operationNameSuffix + responseSuffix),
          void 0,
          ResponseType
        )
      ).name
    );
    const operationParameters = apiGen.resolveArray(operation.parameters);
    const pathItemParameters = apiGen.resolveArray(pathItem.parameters).filter((pp) => !operationParameters.some((op) => op.name === pp.name && op.in === pp.in));
    const parameters = (0, import_generate3.supportDeepObjects)([...pathItemParameters, ...operationParameters]).filter(
      argumentMatches(overrides?.parameterFilter)
    );
    const allNames = parameters.map((p) => p.name);
    const queryArg = {};
    function generateName(name, potentialPrefix) {
      const isPureSnakeCase = /^[a-zA-Z][a-zA-Z0-9_]*$/.test(name);
      const hasNamingConflict = allNames.filter((n) => n === name).length > 1;
      if (hasNamingConflict) {
        name = `${potentialPrefix}_${name}`;
      }
      const camelCaseName = (0, import_lodash.default)(name);
      if (isPureSnakeCase && !allNames.includes(camelCaseName)) {
        name = camelCaseName;
      }
      while (name in queryArg) {
        name = `_${name}`;
      }
      return name;
    }
    for (const param of parameters) {
      const name = generateName(param.name, param.in);
      queryArg[name] = {
        origin: "param",
        name,
        originalName: param.name,
        type: apiGen.getTypeFromSchema((0, import_generate3.isReference)(param) ? param : param.schema, void 0, "writeOnly"),
        required: param.required,
        param
      };
    }
    if (requestBody) {
      const body = apiGen.resolve(requestBody);
      const schema = apiGen.getSchemaFromContent(body.content);
      const type = apiGen.getTypeFromSchema(schema);
      const schemaName = (0, import_lodash.default)(
        type.name || (0, import_generate3.getReferenceName)(schema) || typeof schema === "object" && "title" in schema && schema.title || "body"
      );
      const name = generateName(schemaName in queryArg ? "body" : schemaName, "body");
      queryArg[name] = {
        origin: "body",
        name,
        originalName: schemaName,
        type: apiGen.getTypeFromSchema(schema, void 0, "writeOnly"),
        required: true,
        body
      };
    }
    const propertyName = (name) => {
      if (typeof name === "string") {
        return (0, import_generate3.isValidIdentifier)(name) ? factory.createIdentifier(name) : factory.createStringLiteral(name);
      }
      return name;
    };
    const queryArgValues = Object.values(queryArg);
    const isFlatArg = flattenArg && queryArgValues.length === 1;
    const QueryArg = factory.createTypeReferenceNode(
      registerInterface(
        factory.createTypeAliasDeclaration(
          [factory.createModifier(import_typescript4.default.SyntaxKind.ExportKeyword)],
          capitalize(operationName + operationNameSuffix + argSuffix),
          void 0,
          queryArgValues.length > 0 ? isFlatArg ? withQueryComment(
            factory.createUnionTypeNode([
              queryArgValues[0].type,
              ...!queryArgValues[0].required ? [factory.createKeywordTypeNode(import_typescript4.default.SyntaxKind.UndefinedKeyword)] : []
            ]),
            queryArgValues[0],
            false
          ) : factory.createTypeLiteralNode(
            queryArgValues.map(
              (def) => withQueryComment(
                factory.createPropertySignature(
                  void 0,
                  propertyName(def.name),
                  (0, import_generate3.createQuestionToken)(!def.required),
                  def.type
                ),
                def,
                true
              )
            )
          ) : factory.createKeywordTypeNode(import_typescript4.default.SyntaxKind.VoidKeyword)
        )
      ).name
    );
    const tagOverrides = overrides && (overrides.providesTags !== void 0 || overrides.invalidatesTags !== void 0) ? {
      ...overrides.providesTags !== void 0 ? { providesTags: overrides.providesTags } : {},
      ...overrides.invalidatesTags !== void 0 ? { invalidatesTags: overrides.invalidatesTags } : {}
    } : void 0;
    return generateEndpointDefinition({
      operationName: operationNameSuffix ? capitalize(operationName + operationNameSuffix) : operationName,
      type: isQuery2 ? "query" : "mutation",
      Response: ResponseTypeName,
      QueryArg,
      queryFn: generateQueryFn({
        operationDefinition,
        queryArg,
        isQuery: isQuery2,
        isFlatArg,
        encodePathParams,
        encodeQueryParams
      }),
      extraEndpointsProps: isQuery2 ? generateQueryEndpointProps({ operationDefinition }) : generateMutationEndpointProps({ operationDefinition }),
      tags,
      tagOverrides
    });
  }
  function generateQueryFn({
    operationDefinition,
    queryArg,
    isFlatArg,
    isQuery: isQuery2,
    encodePathParams: encodePathParams2,
    encodeQueryParams: encodeQueryParams2
  }) {
    const { path: path4, verb } = operationDefinition;
    const bodyParameter = Object.values(queryArg).find((def) => def.origin === "body");
    const rootObject = factory.createIdentifier("queryArg");
    function pickParams(paramIn) {
      return Object.values(queryArg).filter((def) => def.origin === "param" && def.param.in === paramIn);
    }
    function createObjectLiteralProperty(parameters, propertyName) {
      if (parameters.length === 0) return void 0;
      const properties = parameters.map((param) => {
        const value = isFlatArg ? rootObject : accessProperty(rootObject, param.name);
        const encodedValue = encodeQueryParams2 && param.param?.in === "query" ? factory.createConditionalExpression(
          factory.createBinaryExpression(value, import_typescript4.default.SyntaxKind.ExclamationEqualsToken, factory.createNull()),
          void 0,
          factory.createCallExpression(factory.createIdentifier("encodeURIComponent"), void 0, [
            factory.createCallExpression(factory.createIdentifier("String"), void 0, [value])
          ]),
          void 0,
          factory.createIdentifier("undefined")
        ) : value;
        return (0, import_generate3.createPropertyAssignment)(param.originalName, encodedValue);
      });
      return factory.createPropertyAssignment(
        factory.createIdentifier(propertyName),
        factory.createObjectLiteralExpression(properties, true)
      );
    }
    return factory.createArrowFunction(
      void 0,
      void 0,
      Object.keys(queryArg).length ? [factory.createParameterDeclaration(void 0, void 0, rootObject, void 0, void 0, void 0)] : [],
      void 0,
      factory.createToken(import_typescript4.default.SyntaxKind.EqualsGreaterThanToken),
      factory.createParenthesizedExpression(
        factory.createObjectLiteralExpression(
          [
            factory.createPropertyAssignment(
              factory.createIdentifier("url"),
              generatePathExpression(path4, pickParams("path"), rootObject, isFlatArg, encodePathParams2)
            ),
            isQuery2 && verb.toUpperCase() === "GET" ? void 0 : factory.createPropertyAssignment(
              factory.createIdentifier("method"),
              factory.createStringLiteral(verb.toUpperCase())
            ),
            bodyParameter === void 0 ? void 0 : factory.createPropertyAssignment(
              factory.createIdentifier("body"),
              isFlatArg ? rootObject : factory.createPropertyAccessExpression(rootObject, factory.createIdentifier(bodyParameter.name))
            ),
            createObjectLiteralProperty(pickParams("cookie"), "cookies"),
            createObjectLiteralProperty(pickParams("header"), "headers"),
            createObjectLiteralProperty(pickParams("query"), "params")
          ].filter(removeUndefined),
          false
        )
      )
    );
  }
  function generateQueryEndpointProps({}) {
    return {};
  }
  function generateMutationEndpointProps({}) {
    return {};
  }
}
function accessProperty(rootObject, propertyName) {
  return (0, import_generate3.isValidIdentifier)(propertyName) ? factory.createPropertyAccessExpression(rootObject, factory.createIdentifier(propertyName)) : factory.createElementAccessExpression(rootObject, factory.createStringLiteral(propertyName));
}
function generatePathExpression(path4, pathParameters, rootObject, isFlatArg, encodePathParams) {
  const expressions = [];
  const head = path4.replace(/\{(.*?)}(.*?)(?=\{|$)/g, (_, expression, literal) => {
    const param = pathParameters.find((p) => p.originalName === expression);
    if (!param) {
      throw new Error(`path parameter ${expression} does not seem to be defined in '${path4}'!`);
    }
    expressions.push([param.name, literal]);
    return "";
  });
  return expressions.length ? factory.createTemplateExpression(
    factory.createTemplateHead(head),
    expressions.map(([prop, literal], index) => {
      const value = isFlatArg ? rootObject : accessProperty(rootObject, prop);
      const encodedValue = encodePathParams ? factory.createCallExpression(factory.createIdentifier("encodeURIComponent"), void 0, [
        factory.createCallExpression(factory.createIdentifier("String"), void 0, [value])
      ]) : value;
      return factory.createTemplateSpan(
        encodedValue,
        index === expressions.length - 1 ? factory.createTemplateTail(literal) : factory.createTemplateMiddle(literal)
      );
    })
  ) : factory.createNoSubstitutionTemplateLiteral(head);
}

// src/index.ts
var require2 = (0, import_node_module.createRequire)(__filename);
async function generateEndpoints(options) {
  const schemaLocation = options.schemaFile;
  const schemaAbsPath = isValidUrl(options.schemaFile) ? options.schemaFile : import_node_path3.default.resolve(process.cwd(), schemaLocation);
  const sourceCode = await enforceOazapftsTsVersion(async () => {
    return generateApi(schemaAbsPath, options);
  });
  const { outputFile, prettierConfigFile } = options;
  if (outputFile) {
    import_node_fs.default.writeFileSync(
      import_node_path3.default.resolve(process.cwd(), outputFile),
      await prettify(outputFile, sourceCode, prettierConfigFile)
    );
  } else {
    return await prettify(null, sourceCode, prettierConfigFile);
  }
}
function parseConfig(fullConfig) {
  const outFiles = [];
  if ("outputFiles" in fullConfig) {
    const { outputFiles, ...commonConfig } = fullConfig;
    for (const [outputFile, specificConfig] of Object.entries(outputFiles)) {
      outFiles.push({
        ...commonConfig,
        ...specificConfig,
        outputFile
      });
    }
  } else {
    outFiles.push(fullConfig);
  }
  return outFiles;
}
function enforceOazapftsTsVersion(cb) {
  const ozTsPath = require2.resolve("typescript", { paths: [require2.resolve("oazapfts")] });
  const tsPath = require2.resolve("typescript");
  const originalEntry = require2.cache[ozTsPath];
  try {
    require2.cache[ozTsPath] = require2.cache[tsPath];
    return cb();
  } finally {
    if (originalEntry) {
      require2.cache[ozTsPath] = originalEntry;
    } else {
      delete require2.cache[ozTsPath];
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  generateEndpoints,
  parseConfig
});
